frontend/src/metabase-lib/lib/queries/NativeQuery.ts

```diff
@@ -333,11 +338,17 @@ export default class NativeQuery extends AtomicQuery {
     return new NativeQuery(this._originalQuestion, datasetQuery);
   }
 
-  // `replaceCardId` updates the query text to reference a different card.
+  // `replaceCardSlug` updates the query text to reference a different card.
   // Template tags are updated as a result of calling `setQueryText`.
-  replaceCardId(oldId, newId) {
-    const re = new RegExp(`{{\\s*#${oldId}\\s*}}`, "g");
-    const newQueryText = this.queryText().replace(re, () => `{{#${newId}}}`);
+  replaceCardSlug(oldId, newCardSlug) {
+    let newQueryText = this.queryText().replace(
+      new RegExp(`{{\\s*#${oldId}\\s*}}`, "g"),
+      `{{#${newCardSlug}}}`,
+    );
+    newQueryText = newQueryText.replace(
+      new RegExp(`{{\\s*#${oldId}-[a-z0-9-]*\\s*}}`, "g"),
+      `{{#${newCardSlug}}}`,
+    );
```

Comment:
I didn't know how to express this in one regex expression, so I thought it would be clearer to express it in two passes.

If the ID of the card is 123, we want to match {{#123}} and {{#123-foo}} but not {{#1234}} or {{#1234-foo}}.